File: backend/service/TransactionService.go
package service

import (
	"budgetapp-backend/model"
)

var transactions = []model.Transaction{}
var nextID = 1

func GetAllTransactions() []model.Transaction {
	return transactions
}

func CreateTransaction(transaction model.Transaction) model.Transaction {
	transaction.ID = nextID
	nextID++
	transactions = append(transactions, transaction)

	return transaction
}

func DeleteTransaction(id int) bool {
	for i, transaction := range transactions {
		if transaction.ID == id {
			transactions = append(transactions[:i], transactions[i+1:]...)
			return true
		}
	}

	return false
}

func UpdateTransaction(id int, updatedTransaction model.Transaction) (model.Transaction, bool) {
	for i, transaction := range transactions {
		if transaction.ID == id {
			updatedTransaction.ID = id
			transactions[i] = updatedTransaction
			return updatedTransaction, true
		}
	}
	return model.Transaction{}, false
}

func GetMonthlyIncome(month string) float64 {
	TotalIncome := 0.0
	for _, transaction := range transactions {
		if transaction.Type == "income" && len(transaction.Date) >= 7 && transaction.Date[:7] == month {
			TotalIncome += transaction.Amount
		}
	}
	return TotalIncome
}

func GetMonthlyExpenses(month string) float64 {
	TotalExpense := 0.0
	for _, transaction := range transactions {
		if transaction.Type == "expense" && len(transaction.Date) >= 7 && transaction.Date[:7] == month {
			TotalExpense += transaction.Amount
		}
	}
	return TotalExpense
}

func GetBalance(month string) float64 {
	TotalBalance := 0.0
	TotalBalance = GetMonthlyIncome(month) - GetMonthlyExpenses(month)
	return TotalBalance
}

File: frontend/src/app/services/transaction.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import { Transaction, TransactionPayload } from '../models/transaction';

@Injectable({
  providedIn: 'root',
})
export class TransactionService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = 'http://localhost:8080';

  getTransactions(): Observable<Transaction[]> {
    return this.http.get<Transaction[]>(`${this.apiUrl}/transactions`);
  }

  createTransaction(payload: TransactionPayload): Observable<Transaction> {
    return this.http.post<Transaction>(`${this.apiUrl}/transactions`, payload);
  }

  deleteTransaction(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/transactions/${id}`);
  }

  updateTransaction(id: number, payload: TransactionPayload): Observable<Transaction> {
    return this.http.put<Transaction>(`${this.apiUrl}/transactions/${id}`, payload);
  }

  getMonthlyIncome(month: string): Observable<{ total_income: number }> {
    return this.http.get<{ total_income: number }>(`${this.apiUrl}/monthly-income`, {
      params: { month },
    });
  }

  getMonthlyExpense(month: string): Observable<{ total_expenses: number }> {
    return this.http.get<{ total_expenses: number }>(`${this.apiUrl}/monthly-expenses`, {
      params: { month },
    });
  }

  getBalance(month: string): Observable<{ balance: number }> {
    return this.http.get<{ balance: number }>(`${this.apiUrl}/balance`, {
      params: { month },
    });
  }
}
